pkg/reconciler/delivery/delivery.go

```diff
@@ -99,53 +108,114 @@ func (c *Reconciler) updateRoute(ctx context.Context, cfg *v1.Configuration) err
 	if err != nil {
 		return err
 	}
+
 	logger.Info("Applying updated Route object")
 	_, err = c.client.ServingV1().Routes(cfg.Namespace).Update(route)
 
+	// when we have latestRevision = true, we know that we don't need to queue future events
+	if *route.Spec.Traffic[0].LatestRevision {
+		logger.Info("Progressive rollout completed!")
+	} else {
+		logger.Infof("Queueing event for %v seconds later", policy.Threshold)
+		c.followup(cfg, time.Duration(policy.Threshold) * time.Second)
+	}
+
 	return err
 }
 
 // isRouteStatusUpToDate determines if the current Route status already matches our desired state
 func isRouteStatusUpToDate(route *v1.Route, newRevName string) bool {
-	for idx := range route.Status.Traffic {
-		if route.Status.Traffic[idx].RevisionName == newRevName {
-			return true
+	// the Route status is up to date if:
+	// 1. the new Revision is listed in the status traffic targets, AND
+	// 2. the status IngressReady condition's last transition time stamp hasn't expired
+	// OR if:
+	// 3. the new Revision is listed in the status traffic targets, AND
+	// 4. the new Revision already reached 100%
+	var nameListed bool = false
+	for _, t := range route.Status.Traffic {
+		if t.RevisionName == newRevName {
+			nameListed = true
+			break
+		}
+	}
+	if !nameListed {
+		return false
+	}
+	if len(route.Status.Traffic) == 1 || *route.Status.Traffic[1].Percent == 100 {
+		return true
+	}
+	for _, c := range route.Status.Conditions {
+		if c.Type != "IngressReady" {
+			continue
+		}
+		if isLastTransitionExpired(c.LastTransitionTime.Inner) {
+			return false
 		}
 	}
 	return false
 }
 
+// isLastTransitionExpired determines if enough time has elapsed since the last transition of the IngressReady condition
+func isLastTransitionExpired(ltt metav1.Time) bool {
```

Comment:
nit: this should take a policy instead of relying on the global policy. Structure the code so that the global access of the policy is minimal, making it easier to refactor when supporting multiple policies and user defined policies.

Reply:
will do